src/hotspot/share/utilities/checkedCast.hpp

```diff
@@ -25,22 +25,177 @@
 #ifndef SHARE_UTILITIES_CHECKEDCAST_HPP
 #define SHARE_UTILITIES_CHECKEDCAST_HPP
 
+#include "metaprogramming/enableIf.hpp"
 #include "utilities/debug.hpp"
+#include "utilities/macros.hpp"
+#include <limits>
+#include <type_traits>
 
-// In many places we've added C-style casts to silence compiler
-// warnings, for example when truncating a size_t to an int when we
-// know the size_t is a small struct. Such casts are risky because
-// they effectively disable useful compiler warnings. We can make our
-// lives safer with this function, which ensures that any cast is
-// reversible without loss of information. It doesn't check
-// everything: it isn't intended to make sure that pointer types are
-// compatible, for example.
-template <typename T2, typename T1>
-constexpr T2 checked_cast(T1 thing) {
-  T2 result = static_cast<T2>(thing);
-  assert(static_cast<T1>(result) == thing, "must be");
-  return result;
+// Implementation support for checked_cast.
+//
+// Because of details of C++ conversion rules, including undefined or
+// implementation-defined behavior in some cases, along with the possibility
+// of tautological comparison warnings from some compilers, the implementation
+// is not as simple as doing the obvious range checks.
+class CheckedCastImpl {
+  template<typename To, typename From, bool permit_tautology>
+  struct CheckSameSign;
+
+  template<typename To, typename From, bool permit_tautology,
+           bool is_narrowing = (sizeof(To) <= sizeof(From))>
+  struct CheckToSigned;
+
+  template<typename To, typename From, bool permit_tautology,
+           bool is_narrowing = (sizeof(To) < sizeof(From))>
+  struct CheckFromSigned;
+
+  template<typename To, typename From>
+  static constexpr bool check_range(From from) {
+    To to_max = std::numeric_limits<To>::max();
+    return from <= static_cast<From>(to_max);
```

Comment:
Don't we need to check that From is big enough to contain to_max?
Are we allowed to use checked_cast for signed to unsigned widening?
    checked_cast<uint64_t, int32_t> // fail if source is negative